diff --git a/core/java/android/os/Process.java b/core/java/android/os/Process.java
index 1ba52fe..5497557 100644
--- a/core/java/android/os/Process.java
+++ b/core/java/android/os/Process.java
@@ -109,6 +109,12 @@ public class Process {
      */
     public static final int VPN_UID = 1016;
 
+     /**
+     * Defines the UID/GID for the FM process.
+     * @hide
+     */
+    public static final int FM_RADIO_UID = 1029;
+
     /**
      * Defines the UID/GID for the NFC service process.
      * @hide
diff --git a/data/etc/platform.xml b/data/etc/platform.xml
index 12f62fe..b0ef834 100644
--- a/data/etc/platform.xml
+++ b/data/etc/platform.xml
@@ -202,5 +202,7 @@
             file="/system/framework/android.test.runner.jar" />
     <library name="javax.obex"
             file="/system/framework/javax.obex.jar"/>
+    <library name="qcom.fmradio"
+            file="/system/framework/qcom.fmradio.jar"/>
 
 </permissions>
diff --git a/services/java/com/android/server/pm/PackageManagerService.java b/services/java/com/android/server/pm/PackageManagerService.java
index 87eb570..332b015 100755
--- a/services/java/com/android/server/pm/PackageManagerService.java
+++ b/services/java/com/android/server/pm/PackageManagerService.java
@@ -192,10 +192,15 @@ public class PackageManagerService extends IPackageManager.Stub {
     private static final boolean DEBUG_APP_DIR_OBSERVER = false;
     private static final boolean DEBUG_VERIFY = false;
 
+    static final boolean MULTIPLE_APPLICATION_UIDS = true;
     private static final int RADIO_UID = Process.PHONE_UID;
     private static final int LOG_UID = Process.LOG_UID;
     private static final int NFC_UID = Process.NFC_UID;
     private static final int BLUETOOTH_UID = Process.BLUETOOTH_UID;
+    private static final int FM_RADIO_UID = Process.FM_RADIO_UID;
+    static final int FIRST_APPLICATION_UID =
+                   Process.FIRST_APPLICATION_UID;
+    static final int MAX_APPLICATION_UIDS = 1000;
 
     private static final boolean GET_CERTIFICATES = true;
 
@@ -1014,6 +1019,10 @@ public class PackageManagerService extends IPackageManager.Stub {
         mSettings.addSharedUserLPw("com.tmobile.thememanager", THEME_MANAGER_GUID, ApplicationInfo.FLAG_SYSTEM);
         mSettings.addSharedUserLPw("android.uid.nfc", NFC_UID, ApplicationInfo.FLAG_SYSTEM);
         mSettings.addSharedUserLPw("android.uid.bluetooth", BLUETOOTH_UID, ApplicationInfo.FLAG_SYSTEM);
+        mSettings.addSharedUserLPw("android.uid.fm_radio",
+                 MULTIPLE_APPLICATION_UIDS
+                         ? FM_RADIO_UID : FIRST_APPLICATION_UID,
+                 ApplicationInfo.FLAG_SYSTEM);
 
         String separateProcesses = SystemProperties.get("debug.separate_processes");
         if (separateProcesses != null && separateProcesses.length() > 0) {
